### ISSUE TITLE
[Bug] nodes of a database(that isn't the default database) that carry more than one label get deleted after the python-driver script has finished

### ISSUE THREAD

**Issue by test123-all** [neo4j/neo4j-python-driver#679]
### Explanation
Contents of database(that isn't the default database) change after the python script has finished.
Note: If the script is run against the default database it works.

- Neo4j version: 4.4.4 (enterprise) from the official enterprise docker image
- Neo4j Mode: Single instance
- Operating system of the server: Ubuntu Server 20.04

- Driver version: Python driver 4.4.0  
- Operating system of my computer: KDE neon User - 5.24 (based on Ubuntu 20.04)

### **Steps to reproduce**
With a script:
1. Load raw data on the database with the python driver and the `session.write_transaction()`-function
2. Refactor the graph with `session.write_transaction()`-functions
3. (Script finishes)
4. (A few seconds later the data gets changed again, a lot of nodes and labels get deleted)

### Expected behavior
The graph shouldn't change after the refactoring is finished.

### Actual behavior
A lot of nodes and labels get deleted.

There are **no** error messages in the log or in the terminal, I run the python script in.

### What I've tried:
Running the script against the default database(the one with the little house icon in the web browser) works **without** problems.

I am using the neo4j enterprise version in the environment of university research. I contacted the german sales team back in November/December 2021 and they gave me access to other tools but not explicitly to the server enterprise version I asked for. Since the enterprise docker image is downloadable without verification from dockerhub I thought it wont be a problem.

So I'm thinking maybe a license file is missing or something like this?

**Comment by test123-all** [comment]
## Important Addendum:
### To the point:
I have another ongoing Issue at https://github.com/neo4j/neo4j-browser/issues/1699 and did some testing in correspondence of both problems and found out that all nodes that have a second label are deleted in the database that is **not** the **default** one.
**Please note:** I uploaded exactly the same data to both databases.

### Deeper explanation:
On the other issue I found out that the number on the left on the following picture is the actual label count of the cypher query(in that example also the label count of the whole database since I returned all nodes of the database). 
**Please also note**: that the used database is called "neo4j" and **is** the **default** one.
![Screenshot_20220310_222705](https://user-images.githubusercontent.com/84460847/157948140-8de53635-32c1-4df2-b7b2-3962f05aedad.png)

</br>

On the next picture you can see the output I get on the "testservergraphprogram" database that is **not** the **default** one.
The nodes with the labels "ProcessVariable", "Measured" and "Simulated" all have the label "Pipeline" too. 

![Screenshot_20220311_210557](https://user-images.githubusercontent.com/84460847/157951554-b484bf08-c3f1-47d9-a065-10759b9d04a1.png)

Knowing this, we can take the numbers from the pictures and do some math:
If we take the label counts of the  "ProcessVariable", "Measured" and "Simulated" labels(of the first picture) and add them together we get: `432 + 1326 + 1191 = 2949`.  2949 is the label count of "Pipeline", so there is nothing missing.

If we further  take the number for the label occurrences of "Pipeline" which is also the amount of nodes that carry the "Pipeline" label and subtract it from the total node count of the whole database, of the first picture, we get `5625 - 2949 = 2676`. 
2676 is **exactly** the total node count of the second picture.

**-->** Therefore I anticipate that the nodes, that carry more than one label, get somehow deleted in a database, that is **not** the **default** one.

**Please also note:** The "deletion" happens **after** the Neo4j-Python-Driver script finishes. So maybe it could be a problem that arises somewhere different than the python driver(I can see all nodes and the right formatting for a little time).

**Comment by robsdedude** [comment]
Hi. Thanks for reaching out.

First of all, no, just because you might technically be able use the enterprise edition, doesn't mean you are legally allowed to.

Now for your issue. What you describe cannot come from the driver. When the program terminates, there is nothing the driver can do. I guess with a periodic commit query or an auto commit transaction you could maybe get some effects after the end of the program using the driver, but a) I never tested this hypothesis, b) you wrote you use `session.write_transaction()` (i.e., managed transactions, a.k.a. transaction functions), with those, this cannot happen.

To help you further debug this, kindly provide a minimal example so that I can reproduce the issue locally.

I suspect either a bug in your script, some issue the server is having (maybe a full disk or so?)-for that check the server's logs, or maybe another program is sending queries and altering the DB's content.

**Comment by robsdedude** [comment]
Hi. Thanks for reaching out.

First of all, no, just because you might technically be able to use the enterprise edition, doesn't mean you are legally allowed to.

Now for your issue. What you describe cannot come from the driver. When the program terminates, there is nothing the driver can do. I guess with a periodic commit query or an auto commit transaction you could maybe get some effects after the end of the program using the driver, but a) I never tested this hypothesis, b) you wrote you use `session.write_transaction()` (i.e., managed transactions, a.k.a. transaction functions), with those, this cannot happen.

To help you further debug this, kindly provide a minimal example so that I can reproduce the issue locally.
Here is a simple script you might want to use as a starting point for building that minimal example.

```python
import neo4j


def query(tx, query_, **params):
    result = tx.run(query_, params)
    return result.data()


def main():
    auth = ('neo4j', 'pass')
    with neo4j.GraphDatabase.driver('bolt://localhost:7687', auth=auth) as d:
        with d.session() as s1:
            s1.write_transaction(query,
                                 "CREATE DATABASE testdb IF NOT EXISTS")
        for database in (None, "testdb"):
            with d.session(database=database,
                           bookmarks=[s1.last_bookmark()]) as s:
                # make sure db is empty
                s.write_transaction(query, "MATCH (n) DETACH DELETE n")
                # create some nodes
                s.write_transaction(query, "CREATE (:TypeA:TypeB)")
                s.write_transaction(query, "CREATE (:TypeA)")
                # TODO: what do I have to do with the data to see the described
                #       effect/bug?
                ...


if __name__ == "__main__":
    main()
```

I suspect either a bug in your script, some issue the server is having (maybe a full disk or so?)-for that, please check the server logs, or maybe another program is sending queries and altering the DB's content.

**Comment by test123-all** [comment]
@robsdedude
Hey,
thanks for answering my issue. 

**About the licensing-problem:**
Like I mentioned above I wrote to the german sales team in November/December 2021. Since I'm not sure anynmore if I can use the enterprise-edition I would like to talk to someone what I need to pay or do to use it. I'm writing my Bachelor-Thesis about processing measurement-data at the Technical University of Darmstadt, Germany.
Can you please provide more information about what I should do next? For example writing the german-sales team again and ask for some kind of written confirmation? 


**To the bug:**
I solved it with your help and you were right! :) 
After your suggestion I double checked if there is a program that runs queries automatically because I had one doing this. And I forgot to comment out the line in the crontab. But that program shouldn't do any refactoring of the graph on its own. Turned out that this part of the code wasn't indented properly(one indention missing) too. That's all.
Basically I tried to upload different data sets to the server and the automatic program formatted the data while it got uploaded.

Since the issue is solved I close it by myself.

Thanks again for your help!

**Comment by jexp** [comment]
@test123-all drop me an email to michael at neo4j.com and we'll sort it out. For just writing your thesis it shouldn't be a big deal.

**Comment by test123-all** [comment]
@jexp Thanks for helping me with this problem.

I sent a message to your email with my official student-email of the university.
